### PR TITLE
Keep correct record of synced references

### LINKED ISSUE #12805: [BUG] Serialization of new models breaks with Document.hold()
### Software versions

Bokeh 3.1.0dev2

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

Serialization should work even if multiple models are serialized consecutively due to a `Document.hold()`.

### Observed behavior

When running the following code we see that, instead of serializing the `ImportedStyleSheet` model entirely the first time, it sent just a reference.

### Example code

```Python
from bokeh.io import curdoc
from bokeh.models import Button, Column, Div, Row, ImportedStyleSheet

b = Button(label='Click')
d1 = Div(text='<b>Some HTML</b>')
d2 = Div(text='<b>Some HTML</b>')

def add(event):
    curdoc().hold()
    d1.stylesheets = [ImportedStyleSheet(url='foo.css')]
    d2.stylesheets = [ImportedStyleSheet(url='bar.css')]
    curdoc().unhold()

b.on_click(add)

column = Column(children=[d1, d2, b])

curdoc().add_root(column)
```

### Stack traceback or browser console output

_No response_

### Screenshots

Screen Shot 2023-02-14 at 10 37 30

### PR DESCRIPTION
**Pull request by mattpap**
fixes #12805

### CODE REVIEW

**Comment by philippjfr** [comment]
Can confirm that this fixes the issues I was seeing even in more complex setups.

**Comment by philippjfr** [comment]
Can we get a dev release once this is merged?

File: tests/unit/bokeh/document/test_document.py [hunk 1]
```diff
@@ -873,45 +873,48 @@ def patch_test(new_value: Any):
         assert 'screen' == root1.foo_units
 
     def test_patch_reference_property(self) -> None:
-        d = document.Document()
-        assert not d.roots
-        assert len(d.models) == 0
```
Comment (bryevdv) [inline tests/unit/bokeh/document/test_document.py]: Is this necessary? There's only one important line change in this commit but that's really not immediately apparent from a page full of diff.

File: tests/unit/bokeh/protocol/messages/test_patch_doc.py [hunk 2]
```diff
@@ -58,6 +58,7 @@ def _sample_doc(self):
         another = AnotherModelInTestPatchDoc()
         doc.add_root(SomeModelInTestPatchDoc(child=another))
         doc.add_root(SomeModelInTestPatchDoc())
+        doc.to_json() # clear new model queue
```
Comment (bryevdv) [inline tests/unit/bokeh/protocol/messages/test_patch_doc.py:61]: It should also be possible to have a unit test of `flush_synced` that directly exercises the new `is_still_new` parameter, I think?

**Review by ianthomas23 — APPROVED** [review]
I'm approving this without really understanding the details as I know we need to cut a dev release containing it so that it can be tested fully.

**Pull request merged**